airdrop-machine/sources/airdrop_machine.move

```diff
@@ -0,0 +1,289 @@
+            royalty::get(collection),
+            uri,
+        );
+
+        token_components::create_refs(constructor_ref);
+        transfer_token::transfer(&collection_signer, recipient_addr, constructor_ref);
+        object::object_from_constructor_ref(constructor_ref)
+    }
+
+    public fun create_collection_impl(
+        admin: &signer,
+        description: String,
+        name: String,
+        uri: String,
+        mutable_collection_metadata: bool,
+        mutable_token_metadata: bool,
+        tokens_burnable_by_collection_owner: bool,
+        tokens_transferrable_by_collection_owner: bool,
+        max_supply: Option<u64>,
+        royalty_numerator: Option<u64>,
+        royalty_denominator: Option<u64>,
+    ): Object<Collection> {
+        let admin_addr = signer::address_of(admin);
+        let object_constructor_ref = &object::create_object(admin_addr);
+        let object_signer = object::generate_signer(object_constructor_ref);
+        let royalty = royalty(&mut royalty_numerator, &mut royalty_denominator, admin_addr);
+
+        let constructor_ref = if (option::is_some(&max_supply)) {
+            collection::create_fixed_collection(
+                &object_signer,
+                description,
+                option::extract(&mut max_supply),
+                name,
+                royalty,
+                uri,
+            )
+        } else {
+            collection::create_unlimited_collection(
+                &object_signer,
+                description,
+                name,
+                royalty,
+                uri,
+            )
+        };
+
+        collection_components::create_refs_and_properties(&constructor_ref);
+        let collection = object::object_from_constructor_ref(&constructor_ref);
+        configure_collection_and_token_properties(
+            &object_signer,
+            collection,
+            mutable_collection_metadata,
+            mutable_token_metadata,
+            tokens_burnable_by_collection_owner,
+            tokens_transferrable_by_collection_owner,
+        );
+
+        let object_signer = object::generate_signer(&constructor_ref);
+
+        move_to(&object_signer, CollectionConfig {
+            extend_ref: object::generate_extend_ref(object_constructor_ref),
+            ready_to_mint: false,
+        });
+
+        collection
+    }
+
+    fun royalty(
+        royalty_numerator: &mut Option<u64>, 
+        royalty_denominator: &mut Option<u64>, 
+        admin_addr: address
+    ): Option<Royalty> {
+        assert!(option::is_some(royalty_numerator) == option::is_some(royalty_denominator), error::invalid_argument(EROYALTY_CONFIGURATION_INVALID));
+        if (option::is_some(royalty_numerator) && option::is_some(royalty_denominator) && option::extract(royalty_numerator) != 0 && option::extract(royalty_denominator) != 0) {
+            option::some(royalty::create(option::extract(royalty_numerator), option::extract(royalty_denominator), admin_addr))
```

Comment:
This is really messy, I think it's best we have the royalty num/dom be u64, just like we do for `aptos_token`.

Reply:
for this use case, royalty is none. so we shouldn't make royalty non-optional. i agree its messy but it provides flexibility.